### PR TITLE
Add test_max_fps for FIM.

### PR DESCRIPTION
**Pull request by antoniomanuelfr**
Hello team,

This PR aims to add a new integration test that checks that the FIM scan sleeps when reached 
the maximum files per second

Closes https://github.com/wazuh/wazuh/issues/5645

# Tests logic

- The test will create a number of files and set a low `max_fps`. Then it will wait that the debug message appears.
- In the second case it will check that if the option is disabled (setting `max_fps` to 0) the debug messages doesn't appear.

# Tests checks

- [X] Proven that tests **pass** when they have to pass
- [X] Proven that tests **fail** when they have to fail
- [ ] Tested and passed in Jenkins. Build URL: `<YOUR_JENKINS_BUILD_URL>`

## RPM manager

```
[root@centosmanager integration]# python3 -m pytest test_fim/test_max_fps/
=========================================================== test session starts ============================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /vagrant/wazuh-qa/tests/integration, configfile: pytest.ini
plugins: metadata-1.11.0, html-2.0.1, testinfra-5.0.0
collected 2 items                                                                                                                          

test_fim/test_max_fps/test_max_fps.py ..                                                                                             [100%]


```

Best regards.

### CODE REVIEW

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import os
+
+import re
+import pytest
+from time import time
+from wazuh_testing import global_parameters
+from wazuh_testing.fim import LOG_FILE_PATH, generate_params, create_file, REGULAR
+from wazuh_testing.tools import PREFIX
+from wazuh_testing.tools.configuration import load_wazuh_configurations, check_apply_test
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:13]: `check_apply_test ` is imported but unused.

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# Copyright (C) 2015-2020, Wazuh Inc.
+# Created by Wazuh, Inc. <info@wazuh.com>.
+# This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+
+import os
+
+import re
+import pytest
+from time import time
+from wazuh_testing import global_parameters
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:10]: Add a blank line to separate system imports from `wazuh_testing` imports.

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    return request.param
+
+
+# Functions
+
+def callback_detect_max_fps(line):
+    msg = r'.*Maximum number of files read per second reached, sleeping.$'
+    match = re.match(msg, line)
+
+    return match is not None
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:51]: This callback should be in _fim.py_.

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
+test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+configurations_path = os.path.join(test_data_path, 'wazuh_conf.yaml')
+
+# Configurations
+
+conf_params = {'TEST_DIRECTORIES': test_directories[0],
+               'MAX_FPS': [MAX_FPS, 0]
+               }
+p, m = generate_params(extra_params=conf_params, modes=['scheduled', 'scheduled'])
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:33]: `scheduled` twice?

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+# Tests
+
+
+def test_max_fps(get_configuration, configure_environment, restart_syscheckd, wait_for_fim_start):
+    """
+    Check that FIM sleeps for one second when the option max_fps is enabled
+
+    Parameters
+    ----------
+
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:67]: Remove this section if there are no parameters.

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+@pytest.fixture(scope='module', params=configurations)
+def get_configuration(request):
+    """Get configurations from the module."""
+    return request.param
+
+
+# Functions
+
+def callback_detect_max_fps(line):
+    msg = r'.*Maximum number of files read per second reached, sleeping.$'
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:48]: The `$` is not needed here. Also, the original message ends in `.` so the last `.` can be substituted with `\.` here.

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    match = re.match(msg, line)
+
+    return match is not None
+
+
+def extra_configuration_before_yield():
+    for _ in range(50):
+        create_file(REGULAR, test_directories[0], f'test_{int(round(time() * 10 ** 6))}', content='')
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:56]: You can have here any number, there is no need to calculate this big number to add it to the file name. You can just use `i` from the `for` loop.

File: tests/integration/test_fim/test_max_fps/test_max_fps.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    Parameters
+    ----------
+
+    """
+    try:
+        wazuh_log_monitor.start(timeout=global_parameters.default_timeout * 2,
+                                callback=callback_detect_max_fps)
+    except TimeoutError:
+        if get_configuration['metadata']['max_fps'] == 0:
+            pass
```
Comment (jesusjimsa) [inline tests/integration/test_fim/test_max_fps/test_max_fps.py:74]: You need to add a case for when the `max_fps` value is not `0`. In this case, if there is a `TimeoutError`, the test should fail.
As it is now the test will always pass.

**Review by jesusjimsa — CHANGES_REQUESTED** [review]

**Review by jesusjimsa — CHANGES_REQUESTED** [review]
The test failed in my computer: 
[report_max_fps.zip](https://github.com/wazuh/wazuh-qa/files/5673959/report_max_fps.zip)

**Comment by antoniomanuelfr** [comment]
> The test failed in my computer:
> [report_max_fps.zip](https://github.com/wazuh/wazuh-qa/files/5673959/report_max_fps.zip)
Hi, the test was failing due to a typo in the main branch that has been fixed.
Could you check it again?

**Review by jesusjimsa — DISMISSED** [review]
LGTM!

**Review by Molter73 — APPROVED** [review]
LGTM!

**Review by Molter73 — DISMISSED** [review]

**Comment by antoniomanuelfr** [comment]
After the last modifications in https://github.com/wazuh/wazuh/pull/6830 (this commit adapts the function that checks the `max_files_per_second` option https://github.com/wazuh/wazuh/pull/6830/commits/d12ff228b9c6326b8495bc73e552fa96959e490b to work with all FIM modes) the integration test was modified to adapt to the new behavior.

This is the report of this test only with the latest changes:
[Uploading report.zip…]()

**Review by Molter73 — APPROVED** [review]
LGTM!

File: tests/integration/test_fim/test_files/test_max_files_per_second/test_max_files_per_second.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# Copyright (C) 2015-2020, Wazuh Inc.
```
Comment (BraulioV) [inline tests/integration/test_fim/test_files/test_max_files_per_second/test_max_files_per_second.py:1]:
```suggestion
# Copyright (C) 2015-2021, Wazuh Inc.
```

File: tests/integration/test_fim/test_files/test_max_files_per_second/test_max_files_per_second.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    scheduled = get_configuration['metadata']['fim_mode'] == 'scheduled'
+    check_time_travel(scheduled, monitor=wazuh_log_monitor)
+    try:
+        wazuh_log_monitor.start(timeout=global_parameters.default_timeout * 2,
+                                callback=callback_detect_max_files_per_second)
+    except TimeoutError:
+        if get_configuration['metadata']['max_files_per_sec'] == 0:
+            pass
+        else:
+            raise TimeoutError
```
Comment (BraulioV) [inline tests/integration/test_fim/test_files/test_max_files_per_second/test_max_files_per_second.py:64]:
```suggestion
    except TimeoutError as e:
        if get_configuration['metadata']['max_files_per_sec'] == 0:
            pass
        else:
            raise e
```

**Review by BraulioV — CHANGES_REQUESTED** [review]

**Review by BraulioV — APPROVED** [review]

**Pull request merged**